### PR TITLE
AI - Improved Mob Jumping

### PR DESCRIPTION
**Pull request by SafwatHalaby**
Mobs now have a decent jump ability, and they go up blocks much more efficiently.

### CODE REVIEW

File: src/Mobs/Monster.h [hunk 1]
```diff
@@ -217,6 +217,7 @@ class cMonster :
 
 	virtual void HandleFalling(void);
 	int m_LastGroundHeight;
+	int m_JumpCoolDown;
```
Comment (tigerw) [inline src/Mobs/Monster.h]: unsigned int?

**Comment by tigerw** [comment]
Still have to fix collision detection though.

File: src/Mobs/Monster.cpp [hunk 2]
```diff
@@ -286,12 +287,18 @@ void cMonster::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
...
 				m_bOnGround = false;
-
+				m_JumpCoolDown = 20;
 				// TODO: Change to AddSpeedY once collision detection is fixed - currently, mobs will go into blocks attempting to jump without a teleport
-				AddPosY(1.2);  // Jump!!
+				AddPosY(1.6);  // Jump!!
+				SetSpeedX(3.2 * (m_Destination.x - GetPosition().x));  // Move forward in a preset speed.
+				SetSpeedZ(3.2 * (m_Destination.z - GetPosition().z));  // The numbers were picked based on trial and error and 1.6 and 3.2 are perfect.
+			}
+			else if (m_JumpCoolDown > 0)
```
Comment (NiLSPACE) [inline src/Mobs/Monster.cpp]: Is this check needed? If it isn't `0` we can be sure it's above it right?

Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: Good question, and no, we can't, with your proposal:
If m_JumpCoolDown is 0 and DoesPosYRequireJump returns 0, we'll get an m_JumpCoolDown of -1.

File: src/Mobs/Monster.h (hunk 1 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.h]: Why?

Comment (SafwatHalaby) [inline src/Mobs/Monster.h]: It's a value which is between 20 and 0, I don't see the merit of an unsigned int here. I could even use an int_8 (perhaps I should?)

File: src/Mobs/Monster.cpp (hunk 2 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: You're right though, the code is confusing, I'll push a little improvement.

File: src/Mobs/Monster.h (hunk 1 above)
Comment (madmaxoft) [inline src/Mobs/Monster.h]: No, use a regular `int` or `unsigned` for normal stuff. Architectures such as ARM prefer ints to bytes.

File: src/Mobs/Monster.cpp (hunk 2 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: Turns out I didn't need the extra check if the code is written a bit differently, done.

File: src/Mobs/Monster.cpp [hunk 3]
```diff
@@ -286,12 +287,21 @@ void cMonster::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
...
-				// TODO: Change to AddSpeedY once collision detection is fixed - currently, mobs will go into blocks attempting to jump without a teleport
-				AddPosY(1.2);  // Jump!!
+				if (DoesPosYRequireJump((int)floor(m_Destination.y)))
+				{
+						m_bOnGround = false;
+						m_JumpCoolDown = 20;
+						// TODO: Change to AddSpeedY once collision detection is fixed - currently, mobs will go into blocks attempting to jump without a teleport
+						AddPosY(1.6);  // Jump!!
+						SetSpeedX(3.2 * (m_Destination.x - GetPosition().x));  // Move forward in a preset speed.
+						SetSpeedZ(3.2 * (m_Destination.z - GetPosition().z));  // The numbers were picked based on trial and error and 1.6 and 3.2 are perfect.
```
Comment (madmaxoft) [inline src/Mobs/Monster.cpp]: Too much indent here.

**Comment by SafwatHalaby** [comment]
Fixed and squashed.

File: src/Mobs/Monster.h (hunk 1 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.h]: No need for unsigned then.

File: src/Mobs/Monster.cpp [hunk 4]
```diff
@@ -286,12 +287,21 @@ void cMonster::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
...
 		if (m_bOnGround)
 		{
-			if (DoesPosYRequireJump((int)floor(m_Destination.y)))
+			if (m_JumpCoolDown == 0)
 			{
-				m_bOnGround = false;
-
-				// TODO: Change to AddSpeedY once collision detection is fixed - currently, mobs will go into blocks attempting to jump without a teleport
-				AddPosY(1.2);  // Jump!!
+				if (DoesPosYRequireJump((int)floor(m_Destination.y)))
```
Comment (madmaxoft) [inline src/Mobs/Monster.cpp]: Use a `static_cast<int>(...)`
We're trying to get rid of the C-style casts.

Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: Done.

**Comment by SafwatHalaby** [comment]
Looks good to me. Squash?

**Comment by madmaxoft** [comment]
Yup

**Comment by SafwatHalaby** [comment]
Squashed, feel free to merge.

**Comment by SafwatHalaby** [comment]
This is offtopic but: Aren't the CI's stupid? Why do they have to build it all from scratch every time? Can't they cache their object files or something?

**Comment by madmaxoft** [comment]
You want them to check that you haven't forgotten any files, or that there aren't any unwanted dependencies.

**Pull request merged**